core/storage/car/cids_index/cids_index.cpp

```diff
@@ -11,82 +11,186 @@
+    if (read(car_file, varint)) {
+      if (common::read(car_file, prefix) && prefix == kCborBlakePrefix) {
+        if (common::read(car_file, key) && key == row.key) {
+          if (end) {
+            *end = row.offset.value() + varint.length + varint.value;
+          }
+          return {true, varint.value - prefix.size() - key.size()};
+        }
+      }
     }
-    return index;
+    return {false, 0};
   }
 
-  outcome::result<std::shared_ptr<Index>> create(const std::string &car_path,
-                                                 const std::string &index_path,
-                                                 IpldPtr ipld,
-                                                 Progress *progress) {
-    boost::system::error_code ec;
-    auto init_car_size{boost::filesystem::file_size(car_path, ec)};
-    if (ec) {
-      return ec;
-    }
-    if (progress) {
-      progress->car_size = init_car_size;
-      progress->begin();
+  RowsInfo &RowsInfo::feed(const Row &row) {
+    valid = valid && !row.isMeta();
+    if (valid) {
+      sorted = count == 0 || max_key < row.key;
+      valid = sorted;
+      ++count;
+      if (row.offset.value() > max_offset.offset.value()) {
+        max_offset = row;
+      }
+      max_key = std::max(max_key, row.key);
     }
-    auto BOOST_OUTCOME_TRY_UNIQUE_NAME{gsl::finally([&] {
-      if (progress) {
-        progress->end();
+    return *this;
+  }
+
+  bool MergeRange::empty() const {
+    return current >= rows.size() && begin >= end;
+  }
+
+  bool MergeRange::read() {
+    assert(!empty());
+    if (current >= rows.size()) {
+      file->seekg(begin * sizeof(Row));
+      rows.resize(std::min(rows.size(), end - begin));
+      if (!common::read(*file, gsl::make_span(rows))) {
+        return false;
       }
-    })};
-    if (init_car_size > (size_t{40} << 30)) {
-      return ERROR_TEXT(
-          "cids_index create: TODO: safe indexing for car over 40gb");
+      begin += rows.size();
+      current = 0;
     }
-    std::ifstream car_file{car_path, std::ios::binary};
-    // estimated
-    car_file.rdbuf()->pubsetbuf(nullptr, 64 << 10);
+    return true;
+  }
 
-    size_t offset{};
-    Buffer header;
-    if (auto varint{codec::uvarint::readBytes(car_file, header)}) {
-      offset += varint + header.size();
-    } else {
-      return ERROR_TEXT("cids_index create: read header failed");
+  void MergeRange::pop() {
+    assert(current < rows.size());
+    ++current;
+  }
+
+  outcome::result<void> merge(std::ostream &out,
+                              std::vector<MergeRange> &&ranges) {
+    auto read_error{ERROR_TEXT("merge: read error")};
+    auto write_error{ERROR_TEXT("merge: write error")};
+    std::greater<MergeRange> cmp;
+    for (auto it{ranges.begin()}; it != ranges.end();) {
+      auto &range{*it};
+      if (range.empty()) {
+        it = ranges.erase(it);
+      } else {
+        // estimated, 64kb
+        range.rows.resize(1638);
```

Comment:
It's not 64 kb

Reply:
`1638*sizeof(Row) = 1638*40 = 65520 = 64kb - 16b`